File: api-app/src/main/java/de/muspellheim/activitysampling/api/application/RecentActivitiesProjection.java
// Copyright (c) 2025 Falko Schumann. All rights reserved. MIT license.

package de.muspellheim.activitysampling.api.application;

import de.muspellheim.activitysampling.api.domain.activities.Activity;
import de.muspellheim.activitysampling.api.domain.activities.RecentActivitiesQuery;
import de.muspellheim.activitysampling.api.domain.activities.RecentActivitiesQueryResult;
import de.muspellheim.activitysampling.api.domain.activities.TimeSummary;
import de.muspellheim.activitysampling.api.domain.activities.WorkingDay;
import de.muspellheim.activitysampling.api.infrastructure.ActivityLoggedEvent;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Stream;

class RecentActivitiesProjection {

  private final LocalDate today;
  private final LocalDate yesterday;
  private final LocalDate thisWeekStart;
  private final LocalDate thisWeekEnd;
  private final LocalDate thisMonthStart;
  private final LocalDate nextMonthStart;
  private final ZoneId timeZone;

  private LocalDate date;
  private final List<Activity> activities = new ArrayList<>();

  private final List<WorkingDay> workingDays = new ArrayList<>();
  private Duration hoursToday = Duration.ZERO;
  private Duration hoursYesterday = Duration.ZERO;
  private Duration hoursThisWeek = Duration.ZERO;
  private Duration hoursThisMonth = Duration.ZERO;

  RecentActivitiesProjection(RecentActivitiesQuery query, Clock clock) {
    timeZone = query.timeZone() != null ? query.timeZone() : ZoneId.systemDefault();
    today = clock.instant().atZone(timeZone).toLocalDate();
    yesterday = today.minusDays(1);
    thisWeekStart = today.minusDays(today.getDayOfWeek().getValue() - 1);
    thisWeekEnd = thisWeekStart.plusDays(6);
    thisMonthStart = today.withDayOfMonth(1);
    nextMonthStart = thisMonthStart.plusMonths(1).withDayOfMonth(1);
  }

  Instant getFrom() {
    return today.minusDays(30).atStartOfDay().atZone(timeZone).toInstant();
  }

  RecentActivitiesQueryResult project(Stream<ActivityLoggedEvent> events) {
    events
        .sorted(Comparator.comparing(ActivityLoggedEvent::timestamp).reversed())
        .map(it -> ActivityMapping.map(it, timeZone))
        .forEach(
            it -> {
              updateWorkingDays(it);
              updateTimeSummary(it);
            });
    createWorkingDay();

    return RecentActivitiesQueryResult.builder()
        .lastActivity(getLastActivity())
        .workingDays(workingDays)
        .timeSummary(getTimeSummary())
        .timeZone(timeZone)
        .build();
  }

  private void updateWorkingDays(Activity activity) {
    var activityDate = activity.dateTime().toLocalDate();
    if (!activityDate.equals(date)) {
      createWorkingDay();
      date = activityDate;
      activities.clear();
    }
    activities.add(activity);
  }

  private void updateTimeSummary(Activity activity) {
    var date = activity.dateTime().toLocalDate();
    var duration = activity.duration();
    if (date.equals(today)) {
      hoursToday = hoursToday.plus(duration);
    }
    if (date.equals(yesterday)) {
      hoursYesterday = hoursYesterday.plus(duration);
    }
    if (!date.isBefore(thisWeekStart) && !date.isAfter(thisWeekEnd)) {
      hoursThisWeek = hoursThisWeek.plus(duration);
    }
    if (!date.isBefore(thisMonthStart) && date.isBefore(nextMonthStart)) {
      hoursThisMonth = hoursThisMonth.plus(duration);
    }
  }

  private void createWorkingDay() {
    if (date == null) {
      return;
    }

    var day = new WorkingDay(date, List.copyOf(activities));
    workingDays.add(day);
  }

  private Activity getLastActivity() {
    if (workingDays.isEmpty()) {
      return null;
    }

    return workingDays.get(0).activities().get(0);
  }

  private TimeSummary getTimeSummary() {
    return TimeSummary.builder()
        .hoursToday(hoursToday)
        .hoursYesterday(hoursYesterday)
        .hoursThisWeek(hoursThisWeek)
        .hoursThisMonth(hoursThisMonth)
        .build();
  }
}

File: src/main/domain/recent_activities_projection.ts
// Copyright (c) 2026 Falko Schumann. All rights reserved. MIT license.

import { Temporal } from "@js-temporal/polyfill";

import {
  isTimestampInPeriod,
  normalizeDuration,
} from "../../shared/domain/temporal";
import {
  type RecentActivitiesQuery,
  RecentActivitiesQueryResult,
  TimeSummary,
  WorkingDay,
} from "../../shared/domain/recent_activities_query";
import { LoggedActivity } from "../../shared/domain/logged_activity";
import type { Projection } from "./projection";
import type { ActivityLoggedEvent } from "./activity_logged_event";

export class RecentActivitiesProjection implements Projection<RecentActivitiesQueryResult> {
  static create({ query }: { query: RecentActivitiesQuery }) {
    return new RecentActivitiesProjection(query);
  }

  readonly #timeZone;
  readonly #from;
  readonly #to;
  readonly #timeSummaryProjection;

  #workingDays: WorkingDay[];
  #date?: Temporal.PlainDate;
  #activities!: LoggedActivity[];

  private constructor(query: RecentActivitiesQuery) {
    this.#timeZone = query.timeZone;
    this.#from = query.today.subtract({ days: 30 });
    this.#to = query.today.with({ day: query.today.daysInMonth });
    this.#timeSummaryProjection = TimeSummaryProjection.create(query);
    this.#workingDays = [];
  }

  update(event: ActivityLoggedEvent) {
    if (
      !isTimestampInPeriod(
        event.timestamp,
        this.#timeZone,
        this.#from,
        this.#to,
      )
    ) {
      return;
    }

    const activityDateTime = event.timestamp
      .toZonedDateTimeISO(this.#timeZone)
      .toPlainDateTime();
    if (
      this.#date == null ||
      !activityDateTime.toPlainDate().equals(this.#date)
    ) {
      this.#createWorkingDay();
      this.#date = activityDateTime.toPlainDate();
      this.#activities = [];
    }
    this.#activities.push(
      LoggedActivity.create({
        dateTime: activityDateTime,
        duration: event.duration,
        client: event.client,
        project: event.project,
        task: event.task,
        notes: event.notes,
        category: event.category,
      }),
    );
    this.#timeSummaryProjection.update(event);
  }

  get(): RecentActivitiesQueryResult {
    this.#createWorkingDay();
    const workingDays = this.#workingDays.reverse();
    const timeSummary = this.#timeSummaryProjection.get();
    return RecentActivitiesQueryResult.create({ workingDays, timeSummary });
  }

  #createWorkingDay() {
    if (this.#date == null) {
      return;
    }

    this.#activities = this.#activities.sort((a1, a2) =>
      Temporal.PlainDateTime.compare(a2.dateTime, a1.dateTime),
    );
    const day = WorkingDay.create({
      date: this.#date,
      activities: this.#activities,
    });
    this.#workingDays.push(day);
  }
}

class TimeSummaryProjection {
  static create({
    today,
    timeZone,
  }: {
    today: Temporal.PlainDate | string;
    timeZone: Temporal.TimeZoneLike;
  }) {
    return new TimeSummaryProjection(today, timeZone);
  }

  readonly #today: Temporal.PlainDate;
  readonly #yesterday: Temporal.PlainDate;
  readonly #weekStart: Temporal.PlainDate;
  readonly #weekEnd: Temporal.PlainDate;
  readonly #monthStart: Temporal.PlainDate;
  readonly #monthEnd: Temporal.PlainDate;
  readonly #timeZone: Temporal.TimeZoneLike;

  #hoursToday;
  #hoursYesterday;
  #hoursThisWeek;
  #hoursThisMonth;

  private constructor(
    today: Temporal.PlainDate | string,
    timeZone: Temporal.TimeZoneLike,
  ) {
    this.#today = Temporal.PlainDate.from(today);
    this.#yesterday = this.#today.subtract("P1D");
    this.#weekStart = this.#today.subtract({ days: this.#today.dayOfWeek - 1 });
    this.#weekEnd = this.#weekStart.add("P6D");
    this.#monthStart = this.#today.with({ day: 1 });
    this.#monthEnd = this.#monthStart.add("P1M").subtract("P1D");
    this.#timeZone = timeZone;
    this.#hoursToday = Temporal.Duration.from("PT0S");
    this.#hoursYesterday = Temporal.Duration.from("PT0S");
    this.#hoursThisWeek = Temporal.Duration.from("PT0S");
    this.#hoursThisMonth = Temporal.Duration.from("PT0S");
  }

  update(event: ActivityLoggedEvent) {
    const date = event.timestamp
      .toZonedDateTimeISO(this.#timeZone)
      .toPlainDate();
    const hours = event.duration;
    if (date.equals(this.#today)) {
      this.#hoursToday = this.#hoursToday.add(hours);
    }
    if (date.equals(this.#yesterday)) {
      this.#hoursYesterday = this.#hoursYesterday.add(hours);
    }
    if (
      Temporal.PlainDate.compare(date, this.#weekStart) >= 0 &&
      Temporal.PlainDate.compare(date, this.#weekEnd) <= 0
    ) {
      this.#hoursThisWeek = this.#hoursThisWeek.add(hours);
    }
    if (
      Temporal.PlainDate.compare(date, this.#monthStart) >= 0 &&
      Temporal.PlainDate.compare(date, this.#monthEnd) <= 0
    ) {
      this.#hoursThisMonth = this.#hoursThisMonth.add(hours);
    }
  }

  get() {
    return TimeSummary.create({
      hoursToday: normalizeDuration(this.#hoursToday),
      hoursYesterday: normalizeDuration(this.#hoursYesterday),
      hoursThisWeek: normalizeDuration(this.#hoursThisWeek),
      hoursThisMonth: normalizeDuration(this.#hoursThisMonth),
    });
  }
}
